<b>Привет, Роман!</b>

Меня зовут Харочкина Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовил замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Таким комментарием выделены удачные решения🙂
</div>

Свои комментарии ты можешь помечать синим цветом.

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так ты можешь оформить свой комментарий
</div>

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

<div class="alert alert-info">
<b>Катя, привет! Спасибо за ревью) постарался все замечания устранить и всё что нужно подправить, жду повторный фидбэк) </b> 
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Давай посмотрим, что получилось)
</div>

<div class="alert alert-info">
<b>Комментарий студента №2</b>
    
<b>Катя, привет ещё раз) Постарался скорректировать оставшийся недочёт, надеюсь в этот раз верно.
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Хорошее начало! С содержанием ориентироваться в проекте намного проще👍 Можно добавить описание данных
</div>

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from IPython.core.display import display, HTML
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score


C:\Users\Games\AppData\Local\Temp\ipykernel_28432\3918809254.py:3: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [2]:
display(HTML("<style>.container { width:88% !important; }</style>"))

In [3]:
try:
    forsen = pd.read_csv('/datasets/insurance.csv')
except:
    forsen = pd.read_csv('C:/Users/Games/Downloads/insurance.csv')

In [4]:
forsen.info(), forsen.shape, display(forsen.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

(None, (5000, 5), None)

In [5]:
print('Дубликатов в датасете:', forsen.duplicated().sum())

Дубликатов в датасете: 153


In [6]:
duplicateRows = forsen[forsen.duplicated()]
display(duplicateRows)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281     1     39.0   48100.0            1                  0
488     1     24.0   32900.0            1                  0
513     0     31.0   37400.0            2                  0
718     1     22.0   32600.0            1                  0
785     0     20.0   35800.0            0                  0
...   ...      ...       ...          ...                ...
4793    1     24.0   37800.0            0                  0
4902    1     35.0   38700.0            1                  0
4935    1     19.0   32700.0            0                  0
4945    1     21.0   45800.0            0                  0
4965    0     22.0   40100.0            1                  0

[153 rows x 5 columns]

In [7]:
forsen = forsen.drop_duplicates()

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Так как у нас нет id клиента и у нас мало столбцов и они достаточно общие, поэтому вполне возможно, что это не дубликаты, а просто совпадающие данные.
</div>

<div class="alert alert-info">
<b>Согласен, вполне возможно, но ввиду их малочисленности, думаю, что ничего страшного, если их убрать. </b> 
</div>

In [8]:
forsen.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                4847 non-null   int64  
 1   Возраст            4847 non-null   float64
 2   Зарплата           4847 non-null   float64
 3   Члены семьи        4847 non-null   int64  
 4   Страховые выплаты  4847 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 227.2 KB


In [9]:
X = forsen.drop(columns='Страховые выплаты', axis=1)
y = forsen['Страховые выплаты']

X.shape, y.shape

((4847, 4), (4847,))

Открыли файл, посмотрели, что в датасете 5к строк, 5 столбцов. Пропусков нет, было 153 дубликата, я их убрал. Не должно сказаться на дальнейших результатах работы. Признаки определили выше, целевой в том числе.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Отлично, данные загружены и изучены! Можно еще посмотреть на статистики по столбцам `describe()`
</div>

<div class="alert alert-info">
<b>Хорошо, буду иметь ввиду, спасибо) </b> 
</div>

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ: b (не изменится)**

**Обоснование: (ниже)**

Начнём с формулы обучения и заменим в ней X на XZ (умножение на обратимую матрицу) и далее раскроем скобки и приведем формулу к такому виду, чтобы можно было показать отношение между $w$ и $w_{new}$.

$$
a_{new} = XZw_{new}
$$

Представим новую матрицу признаков $X1$ как произведение старой $X$ на матрицу $Z$(обратимую):

$$
X1 = X * Z
$$

Подставим новое значение X1 в формулу $w_{new}$:

$$
w_{new} = ((XZ)^T XZ)^{-1} (XZ)^T y
$$

Раскроем первое произведение $(XZ)^T$:

$$
w_{new} = (Z^T X^T XZ)^{-1} Z^T X^T y 
$$

Перегруппируем множители в скобках:

$$
w_{new} = (Z^T (X^T X) Z)^{-1} Z^T X^T y
$$

Так как $Z$ по условию обратимая, то произведение $(Z^T)^{-1} Z^T = E$ (единичная матрица):

$$
w_{new} = Z^{-1} (X^T X)^{-1} E X^T y = Z^{-1} (X^T X)^{-1} X^T y 
$$

Можно заметить, что справа у нас оказалась формула для $w$:

$$
w_{new} = Z^{-1} (X^T X)^{-1} X^T y = Z^{-1} w
$$

Подставим новое значение весов 𝑤1 в формулу для предсказаний линейной регрессии:

$$
a_{new} = XZZ^{-1} w
$$

Так как $Z$ по условию обратимая, то произведение $ZZ^{-1}$ снова равно $E$:

$$
a_{new} = XZZ^{-1} w = XEw = Xw = a
$$

Предсказания $a_{new}$ для матрицы признаков, умноженных на обратимую матрицу $Z$ равны предсказаниям $a$. Качество лин. регрессии при умножении признаков на обратимую матрицу не изменится. Что касается связи для $w$ и $w_{new}$, то $w_{new} = Z^{-1} w$. 

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
По заданию также нужно указать, как связаны параметры линейной регрессии в исходной задаче и в преобразованной. То есть связь между исходными весами $w$ и преобразованными $w_{new}$
</div>

<div class="alert alert-info">
<b>Разве это не подразумевается тем, что мы получаем равенство между предсказаниями? $a_{new} = a = Xw = XZw_{new}$ Параметры ($w|w_{new}$) на которые умножаются признаки ($X|XZ$) равны $w = w_{new}$. Мб что-то не так понял, прояснишь, пожалуйста, как нужно дополнительно показать эту связь? 
</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера №2</b> ❌
    
$X$ не равен $XZ$, а $w$ не равен $w_{new}$. Чтобы получить связь между весами, нужно преобразовать формулу $w_{new} = ((XZ)^T XZ)^{-1} (XZ)^T y$
</div>

<div class="alert alert-info">
    <b>Комментарий студента №2</b>
    
<b>Подзапутался, понял, про параметры и признаки чушь написал. Выше в формулу внёс корректировки при пересмотрении + ниже про преобразование формулы пошагово.
</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера №3</b> ❌
    
Нужно учесть, что для неквадратных матриц  $X$  и  $X^{T}$  обратных матриц не существует, поэтому записывать $X^{-1}$ и $(X^T)^{-1}$ нельзя. Попробуй по-другому раскрыть скобки. Раскрывать  $(X^TX)$ не нужно, так как это квадратная матрица
    
Оставлю полезную ссылку на теорию по обратным матрицам: https://www.berdov.com/works/matrix/obratnaya-matrica/.
    
Нужно добавить уравнение связи для $w$ и $w_{new}$. Это уравнение показывает, как одна переменная выражается через другую 
</div>

<div class="alert alert-info">
    <b>Комментарий студента №3</b>
    
<b>Поменял выше над первым твоим комментом по этому моменту, посмотри, пожалуйста.
</div>

## Алгоритм преобразования

**Алгоритм**

Для шифрования данных будем умножать матрицу признаков на обратимую матрицу, которая будет генерироваться случайным образом.

Этапы алгоритма:

* Составление матрицы;
* проверка матрицы на обратимость;
* получение матрицы преобразованных признаков;
* применение алгоритма на преобразованных признаках и сравнение полученного R2 score с R2 score на оригинальных данных;


<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
В алгоритме нужно описать словами свои дальнейшие действия
</div>

<div class="alert alert-info">
<b>Выше попытался описать алгоритм с учеттом последнего твоего красного замечания, надеюсь сойдет) 
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
👍
</div>

**Обоснование**

Выше обосновано.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Выше мы решили задачу в общем виде. Здесь нужно обосновать то, что конкретный твой алгоритм для нашей конкретной задачи не противоречит общей задаче выше. Обоснование может звучать примерно так: "При умножении признаков на обратимую матрицу качество линейной регрессии не изменится, то есть данный алгоритм позволяет защитить данные, а потом восстановить их, зная обратную матрицу P, если это необходимо"
</div>

<div class="alert alert-info">
<b>Понял, спасибо! 
</div>

## Проверка алгоритма

In [10]:
def data_encrypted(X, y):
    
    Z = np.random.rand(X.shape[1], X.shape[1])
    while np.linalg.det(Z) == 0:
        Z = np.random.rand(X.shape[1], X.shape[1])
    X_encrypted = X@Z
    
    model = LinearRegression()
    model.fit(X_encrypted, y)
    
    y_pred_encrypted = model.predict(X_encrypted)
    r2_encrypted = r2_score(y, y_pred_encrypted)
    
    return X_encrypted, r2_encrypted


X_encrypted, r2_encrypted = data_encrypted(X, y)


model_original = LinearRegression()
model_original.fit(X, y)

y_pred_original = model_original.predict(X)
r2_original = r2_score(y, y_pred_original)

print("Original data R2 score:", r2_original)
print("Encrypted data R2 score:", r2_encrypted)

Original data R2 score: 0.4302010044852067
Encrypted data R2 score: 0.4302010044852015


<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Нужно проверить сгенерированную матрицу на обратимость
</div>

<div class="alert alert-info">
<b>Готово)  
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
👍
</div>

**Метрики одинаковы. Качество линейной регрессии из sklearn не отличается до и после преобразования.**

<div class="alert alert-block alert-success">
    
**Комментарий ревьюера №1** ✅

Верно, качество предсказания действительно не изменяется. Незначительное различие вызвано особенностями преобразования матриц и чисел с плавающей точкой. В случае домножения матрицы признаков на случайную обратимую матрицу можно обеспечить защиту данных. Можно также попробовать декодировать данные обратно.
</div>

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
Ты хорошо постарался, но проект еще нуждается в доработке. Нужно добавить связь между параметрами, описание алгоритма и проверку сгенерированной матрицы на обратимость. Доделывай и присылай на повторную проверку.
        
Жду твоих исправлений 😊
</font></b>
</div>